### DISCUSSION TITLE
"permission denied for schema" Error for Non-Public Schema Connection

### DISCUSSION THREAD

**Discussion by utgarda** [supabase/supabase#20241] [Questions] [answered]
Running into a roadblock trying to connect to a non-public schema in Supabase using supabase-js. Despite following the usual setup steps, I'm hitting a permission error.

### Error:
```js
{
  code: '42501',
  details: null,
  hint: null,
  message: 'permission denied for schema my_nonpublic_schema'
}
```

### What I've Done:
- Exposed the non-public schema.
- Connected using the secret `service_role` key.
- Set up RLS policies for all tables, allowing full access to `service_role`.
- Specified the non-public schema in the connection options.
- Even tried disabling RLS for all tables as a test.

Still, no luck – I keep getting denied. If anyone has any tips or has faced something similar, I'd really appreciate the help!
Should I consider this a bug and report it in GitHub issues?

```typescript
import {
    createClient,
    SupabaseClient,
    SupabaseClientOptions,
} from '@supabase/supabase-js'
import { Database } from './database.types'

export let supabase: SupabaseClient<Database>
...
    const options: SupabaseClientOptions<'swirlbot_dev'> = {
        db: { schema: 'swirlbot_dev' },
    }
    supabase = createClient<Database>(SUPABASE_URL, SUPABASE_KEY, options)
```

@supabase/supabase-js@2.39.2
nodejs: v18.18.2

**Comment by GaryAustin1** [comment, marked as answer]
Did you add grants for the API roles to your schema?
https://supabase.com/docs/guides/api/using-custom-schemas

Note RLS would not apply at all to Service Role but grants to roles do.

**Reply by utgarda** [reply]
Thanks, I granted `USAGE` on the schema and `ALL`  on tables in the schema to `service_role`, and it worked! Now that I know how to use SQL Editor in Supabase Studio, it's time to create a separate role with more granular permissions.

```sql
grant USAGE on schema my_schema to service_role;
grant all on all tables in schema my_schema to service_role;
```

**Reply by ashaller2017** [reply]
Thank you was just encountering the same problem!

**Reply by SaddamMohsen** [reply]
Thank you i have the same issue

**Reply by AshishRam7** [reply]
Thanks a lot! this worked.

**Reply by kingzttpy** [reply]
thank a lot by Demon

**Comment by Djarma12** [comment]
With the help of the Supabase assistant, I solved the same problem, where the data is displayed

![image](https://github.com/user-attachments/assets/f2ac7950-52cc-4362-a3a5-6a6336b1b6f6)
